snappy/click.go

```diff
@@ -481,6 +483,129 @@ func removePackageBinaries(m *snapYaml, baseDir string) error {
 	return nil
 }
 
+// this is a bit confusing, a localestring in xdg is:
+// "key=value" or "key[locale]=" with locale as:
+//  lang_COUNTRY@MODFIER (or a subset of this)
+var desktopFileI18nPattern = `(|\[[a-zA-Z_@]+\])`
+var validDesktopFileLines = []*regexp.Regexp{
+	// headers
+	regexp.MustCompile(`^\[Desktop Entry\]$`),
+	regexp.MustCompile(`^\[Desktop Action`),
+	// whitespace lines
+	regexp.MustCompile(`^\s*$`),
+	// lines with comments
+	regexp.MustCompile(`^\s*#`),
+	// https://specifications.freedesktop.org/desktop-entry-spec/latest/ar01s05.html
+	regexp.MustCompile(`^Type=`),
+	regexp.MustCompile(`^Version=`),
+	regexp.MustCompile(fmt.Sprintf(`^Name%s=`, desktopFileI18nPattern)),
+	regexp.MustCompile(fmt.Sprintf(`^GenericName%s=`, desktopFileI18nPattern)),
+	regexp.MustCompile(`^NoDisplay=`),
+	regexp.MustCompile(fmt.Sprintf(`^Comment%s=`, desktopFileI18nPattern)),
+	regexp.MustCompile(`^Icon=`),
+	regexp.MustCompile(`^Hidden=`),
+	regexp.MustCompile(`^OnlyShowIn=`),
+	regexp.MustCompile(`^NotShowIn=`),
+	regexp.MustCompile(`^Exec=`),
+	regexp.MustCompile(`^TryExec=`),
+	regexp.MustCompile(`^Terminal=`),
+	regexp.MustCompile(`^Actions=`),
+	regexp.MustCompile(`^MimeType=`),
+	regexp.MustCompile(`^Categories=`),
+	regexp.MustCompile(fmt.Sprintf(`^Keywords%s=`, desktopFileI18nPattern)),
+	regexp.MustCompile(`^StartupNotify=`),
+	regexp.MustCompile(`^StartupWMClass`),
+}
+
+func isValidDesktopFilePrefix(line string) bool {
+	for _, re := range validDesktopFileLines {
+		if re.MatchString(line) {
+			return true
+		}
+	}
+	return false
+}
+
+func validExecLine(m *snapYaml, line string) bool {
+	if !strings.HasPrefix(line, "Exec=") && !strings.HasPrefix(line, "TryExec=") {
+		return true
+	}
+	cmd := strings.SplitN(line, "=", 2)[1]
+	for _, app := range m.Apps {
+		validCmd := filepath.Base(generateBinaryName(m, app))
+		// just check the prefix to allow %flag style args
+		// this is ok because desktop files are not run through sh
+		if strings.HasPrefix(cmd, validCmd) {
```

Comment:
Should we make this path absolute when copying it over?

Also, the test should likely be `cmd == validCmd || HasPrefix(cmd, validCmd+" ") 

Not sure TryExec makes much sense in this context. Does it?